src/zarr/storage/_obstore.py

```diff
@@ -0,0 +1,419 @@
+        return f"object://{self.store}"
+
+    def __repr__(self) -> str:
+        return f"ObjectStore({self})"
+
+    def __getstate__(self) -> dict[Any, Any]:
+        state = self.__dict__.copy()
+        state["store"] = pickle.dumps(self.store)
+        return state
+
+    def __setstate__(self, state: dict[Any, Any]) -> None:
+        state["store"] = pickle.loads(state["store"])
+        self.__dict__.update(state)
+
+    async def get(
+        self, key: str, prototype: BufferPrototype, byte_range: ByteRequest | None = None
+    ) -> Buffer | None:
+        # docstring inherited
+        import obstore as obs
+
+        try:
+            if byte_range is None:
+                resp = await obs.get_async(self.store, key)
+                return prototype.buffer.from_bytes(await resp.bytes_async())  # type: ignore[arg-type]
+            elif isinstance(byte_range, RangeByteRequest):
+                bytes = await obs.get_range_async(
+                    self.store, key, start=byte_range.start, end=byte_range.end
+                )
+                return prototype.buffer.from_bytes(bytes)  # type: ignore[arg-type]
+            elif isinstance(byte_range, OffsetByteRequest):
+                resp = await obs.get_async(
+                    self.store, key, options={"range": {"offset": byte_range.offset}}
+                )
+                return prototype.buffer.from_bytes(await resp.bytes_async())  # type: ignore[arg-type]
+            elif isinstance(byte_range, SuffixByteRequest):
+                resp = await obs.get_async(
+                    self.store, key, options={"range": {"suffix": byte_range.suffix}}
+                )
+                return prototype.buffer.from_bytes(await resp.bytes_async())  # type: ignore[arg-type]
+            else:
+                raise ValueError(f"Unexpected byte_range, got {byte_range}")
+        except _ALLOWED_EXCEPTIONS:
+            return None
+
+    async def get_partial_values(
+        self,
+        prototype: BufferPrototype,
+        key_ranges: Iterable[tuple[str, ByteRequest | None]],
+    ) -> list[Buffer | None]:
+        # docstring inherited
+        return await _get_partial_values(self.store, prototype=prototype, key_ranges=key_ranges)
+
+    async def exists(self, key: str) -> bool:
+        # docstring inherited
+        import obstore as obs
+
+        try:
+            await obs.head_async(self.store, key)
+        except FileNotFoundError:
+            return False
+        else:
+            return True
+
+    @property
+    def supports_writes(self) -> bool:
+        # docstring inherited
+        return True
+
+    async def set(self, key: str, value: Buffer) -> None:
+        # docstring inherited
+        import obstore as obs
+
+        self._check_writable()
+        if not isinstance(value, Buffer):
+            raise TypeError(
+                f"ObjectStore.set(): `value` must be a Buffer instance. Got an instance of {type(value)} instead."
+            )
```

Comment:
Any reason to do this check here, but not in `self.set_if_not_exists`? (I'd prefer to remove it)

Reply:
Removed in [`50fd2b9` (#1661)](https://github.com/zarr-developers/zarr-python/pull/1661/commits/50fd2b9010a590ee89ba55a816aa44235b967c81)